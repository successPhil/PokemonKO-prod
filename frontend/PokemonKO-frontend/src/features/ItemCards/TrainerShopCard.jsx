import CardContent from '@mui/material/CardContent';
import Paper from '@mui/material/Paper';
import CardBaloo from '../../styles/CardBaloo';
import Button from '@mui/material/Button';
import TrainerContext from '../../contexts/TrainerContext';
import { useContext, useState} from 'react';
import { Select, InputLabel, MenuItem, FormControl } from '@mui/material';
import { useNavigate } from 'react-router-dom';
import { useEffect } from 'react';
import { makeTransaction } from '../../api/authApi';


export default function TrainerShopCard({item} ) {
    const {trainer, trainerShop, setTrainerShop } = useContext(TrainerContext)
    const [shouldRedirect, setShouldRedirect] = useState(false)
    const [ itemsBought, setItemsBought ] = useState(1)
    const [open, setOpen] = useState(false);
    const navigate = useNavigate()

    useEffect(()=>{
        if (shouldRedirect){
            navigate("/items")
        }
    }, [shouldRedirect])

   

    const handleChange = (event) => {
    setItemsBought(event.target.value);
    };

    const handleClose = () => {
      setOpen(false);
     };

    const handleOpen = () => {
    setOpen(true);
    };

    const updateShopItems = async () => {
        if (trainer.money >= item.value * itemsBought) {
            const updatedItem = trainerShop.find(item => item.id === item.id);
            if (updatedItem){
                updatedItem.quantity -= itemsBought;
                const updatedShopItems = trainerShop.map(item =>
                    item.id === updatedItem.id ? updatedItem : item);
                makeTransaction(item, itemsBought, 'buy')
                setTrainerShop(updatedShopItems)
                setShouldRedirect(true)
            }
        }
    }

    return(
        <>
        <Paper elevation={3} sx={{mt: 8, pt:2, pl:4, pr: 4, height: 100, borderBottomLeftRadius: 15, borderBottomRightRadius: 15, borderTopLeftRadius:15, borderTopRightRadius:15}}>
            <CardContent sx={{display:'flex', justifyContent:'space-between'}}>
                <CardBaloo>{item.name}</CardBaloo>
                <CardBaloo>Quantity: {item.quantity}</CardBaloo>
                <CardBaloo>Value: {item.value}</CardBaloo>
                <FormControl>
                <InputLabel id="shop-controlled-open-filled-select-label"><span className='pokemon-type-input'>Select Amount</span></InputLabel>
                <Select
                labelId="shop-controlled-open-select-label"
                id="shop-controlled-open-select"
                open={open}
                onClose={handleClose}
                onOpen={handleOpen}
                value={itemsBought}
                onChange={handleChange}
                variant='standard'
                >
{[...Array(item.quantity).keys()].map(i => (
        <MenuItem value={i + 1} key={`amountItem${i+1}${i + 1}`}><span className='item-button-text'>{i + 1}</span></MenuItem>
    ))}
                </Select>
                </FormControl>
                <Button variant='contained' color='primary' onClick={updateShopItems}><span className='item-button-text'>Buy Item</span></Button>
            </CardContent>
        </Paper>

        </>
    )
}