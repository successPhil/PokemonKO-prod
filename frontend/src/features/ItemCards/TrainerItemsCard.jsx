import CardContent from '@mui/material/CardContent';
import Paper from '@mui/material/Paper';
import CardBaloo from '../../styles/CardBaloo';
import Button from '@mui/material/Button';
import TrainerContext from '../../contexts/TrainerContext';
import { updateItems } from '../../api/authApi';
import { useContext, useState} from 'react';
import { Select, InputLabel, MenuItem, FormControl } from '@mui/material';
import { useNavigate } from 'react-router-dom';
import { useEffect } from 'react';
import { capitalizeFirst } from '../Enemy/EnemyData';


export default function TrainerItemsCard({ item }) {
    const { endTrainerTurn, selectPokemon, setSelectPokemon, trainerItems, setTrainerItems,  setTrainerDialogue } = useContext(TrainerContext)
    const [ shouldRedirectBattle, setShouldRedirectBattle ] = useState(false)
    const [open, setOpen] = useState(false);
    const [ itemsUsed, setItemsUsed ] = useState(1)
    const navigate = useNavigate()

    useEffect(()=>{
        if (shouldRedirectBattle){
            navigate("/battle")
        }
    }, [shouldRedirectBattle])

    const handleChange = (event) => {
    setItemsUsed(event.target.value);
    };

    const handleClose = () => {
      setOpen(false);
     };

    const handleOpen = () => {
    setOpen(true);
    };


    const updateTrainerItems = async () => {
        // Find the item to update
        const updatedItem = trainerItems.find(item => item.id === item.id);
    
        if (updatedItem && selectPokemon) {
            // Update the quantity
            updatedItem.quantity -= itemsUsed;
            // Create a new array with the updated item
            const updatedTrainerItems = trainerItems.map(item =>
                item.id === updatedItem.id ? updatedItem : item
            );
            // Set the new state
            setTrainerItems(updatedTrainerItems);
            // Send the update to the backend
            updateItems(item.id, itemsUsed);
            console.log(item.item_class)
            if (item.item_class === 'health') {
                const healingAmount = item.stat_boost * itemsUsed
                const missingHealth = selectPokemon.max_health - selectPokemon.health
                const usedHealingAmount = healingAmount > missingHealth ? missingHealth : healingAmount
                const newHealth = selectPokemon.health + healingAmount
                const maxHealth = selectPokemon.max_health
                const updatedHealth = newHealth > maxHealth ? maxHealth : newHealth;
                const itemMsg = `${capitalizeFirst(selectPokemon.name)} has healed ${usedHealingAmount} from using ${itemsUsed} ${item.name}${itemsUsed > 1 ? "s" : ""}! ${capitalizeFirst(selectPokemon.name)} now has ${updatedHealth} health!!!`
                setTrainerDialogue(itemMsg)
                setSelectPokemon((prev) => ({...prev, health: updatedHealth, }))
            }
            if (item.item_class === 'maxhealth'){
                console.log('IN ITEM CLASS CONDITIONAL MAX HEALTH')
                const maxHealthToAdd = item.stat_boost * itemsUsed
                const newMaxHealth = selectPokemon.max_health + maxHealthToAdd
                console.log(newMaxHealth)
                const itemMsg = `${capitalizeFirst(selectPokemon.name)} has gained ${maxHealthToAdd} max health from using ${itemsUsed} ${item.name}${itemsUsed > 1 ? "s" : ""}`
                setTrainerDialogue(itemMsg)
                setSelectPokemon((prev) => ({...prev, max_health: newMaxHealth}))
            }
            if (item.item_class === 'damage'){
                console.log('IN ITEM CLASS CONDITIONAL DAMAGE')
                const powerToAdd = item.stat_boost * itemsUsed
                const newPower = selectPokemon.power + powerToAdd
                const itemMsg = `${capitalizeFirst(selectPokemon.name)} has gained ${powerToAdd} power! from using ${itemsUsed} ${item.name}${itemsUsed > 1 ? "s" : ""}`
                setTrainerDialogue(itemMsg)
                setSelectPokemon((prev) => ({...prev, power: newPower}))
            }
            if (item.item_class === 'defense'){
                console.log('IN ITEM CLASS CONDITIONAL Defense')
                const defenseToAdd = item.stat_boost * itemsUsed
                const newDefense = selectPokemon.defense + defenseToAdd
                const itemMsg = `${capitalizeFirst(selectPokemon.name)} has gained ${defenseToAdd} defense! from using ${itemsUsed} ${item.name}${itemsUsed > 1 ? "s" : ""}`
                setTrainerDialogue(itemMsg)
                setSelectPokemon((prev) => ({...prev, defense: newDefense}))
            }
            endTrainerTurn()
            setShouldRedirectBattle(true)
        }
    };


    return(
        <>
        <Paper elevation={3} sx={{mt: 8, pt:2, pl:4, pr: 4, height: 100, borderBottomLeftRadius: 15, borderBottomRightRadius: 15, borderTopLeftRadius:15, borderTopRightRadius:15}}>
            <CardContent sx={{display:'flex', justifyContent:'space-between'}}>
                <CardBaloo>{item.name}</CardBaloo>
                <CardBaloo>Quantity: {item.quantity}</CardBaloo>
                <CardBaloo>Value: {item.value}</CardBaloo>
                <FormControl>
                <InputLabel id="items-controlled-open-filled-select-label"><span className='pokemon-type-input'>Select Amount</span></InputLabel>
                <Select
                labelId="items-controlled-open-select-label"
                id="items-controlled-open-select"
                open={open}
                onClose={handleClose}
                onOpen={handleOpen}
                value={itemsUsed}
                onChange={handleChange}
                variant='standard'
                >
{[...Array(item.quantity).keys()].map(i => (
        <MenuItem value={i + 1} key={`amountItem${item.id}${i + 1}`}><span className='item-button-text'>{i + 1}</span></MenuItem>
    ))}
                </Select>
                </FormControl>
                {selectPokemon && <Button variant='contained' color='primary' onClick={updateTrainerItems}><span className='item-button-text'>Use Item</span></Button>}
            </CardContent>
        </Paper>

        </>
    )
}