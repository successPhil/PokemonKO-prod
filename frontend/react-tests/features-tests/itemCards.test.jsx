import TrainerItemsCard from '../../src/features/ItemCards/TrainerItemsCard'
import TrainerShopCard from '../../src/features/ItemCards/TrainerShopCard'
import {expect, test} from 'vitest'

test("Trainer Item and Shop Cards render without crashing", ()=>{
    const trainerItemCard = <TrainerItemsCard/>
    expect(trainerItemCard).toMatchSnapshot()
    const trainerShopCard = <TrainerShopCard/>
    expect(trainerShopCard).toMatchSnapshot()
})