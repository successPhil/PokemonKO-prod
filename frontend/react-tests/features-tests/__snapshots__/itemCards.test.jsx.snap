// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Trainer Item and Shop Cards render without crashing 1`] = `<TrainerItemsCard />`;

exports[`Trainer Item and Shop Cards render without crashing 2`] = `<TrainerShopCard />`;
